lib/activerecord/delay_touching/state.rb

```diff
@@ -49,6 +49,18 @@ def clear_records
         @already_updated_records.clear
       end
 
+      # If we don't do this then an infinite loop is possible due to how 
+      # Set#subtract and ActiveRecord::Core#== work internally, ActiveRecord lazily syncs
+      # transaction state after rollback, which may change in-memory state of key objects,
+      # which requires that the hash be rekeyed.
+      def remove_unpersisted_records!
+        @records.each do |attr, set|
+          set.each(&:persisted?)
+          set.instance_variable_get(:@hash).rehash
```

Comment:
I'm not super comfortable with this access into private implementation of Set. 

It looks like Ruby 2.5 introduced the `reset` method to handle exactly this situation (see https://bugs.ruby-lang.org/issues/6589) but that doesn't help us in any prior versions :(.

I'm thinking through whether there's a reasonable alternative to this approach. Otherwise, we could potentially introduce a fallback to this mechanism only if `reset` is not defined (thus using public api for ruby 2.5+ and reducing the risk). I'm definitely open to other suggestions, as well.

Reply:
@mtuckergd @krainboltgreene depending on how far you want to take this, you could just add a private shim to the State class.

```
private

# Set#reset is not introduced until Ruby 2.5
def reset_set(set)
if set.respond_to?(:reset)
   set.reset
else
   set.instance_variable_get(:@hash).rehash
end
```

Reply:
Yep, good call. That's along the lines of the fallback I was considering. I have one other thought I want to pursue before we go this route.

Reply:
Sounds great.